### ISSUE TITLE
sklearn-clj failing when used by ml/evaluate-pipelines

### ISSUE THREAD

**Issue by ash-mcc** [scicloj/scicloj.ml#10]
#### Issue background

Link to the discusion in Zulip: https://clojurians.zulipchat.com/#narrow/stream/151924-data-science/topic/sklearn-clj.20in.20a.20pipeline.20that.20ml.2Fevaluate's.20a.20search-grid.3F 

Software versions used: `scicloj/scicloj.ml 0.2.1` and `scicloj/sklearn-clj 0.3.7`. 

#### Issue summary

The `ml/evaluate-pipelines` example in the first half of the [metamorph.ml README file](https://github.com/scicloj/metamorph.ml) , runs successfully. It uses the `:model-type` `:smile.classification/random-forest`.

If I run the exactly the same code **except** with the `:model-type` `:sklearn.classification/random-forest-classifier` then it errors:
```
Execution error (IllegalArgumentException) at tech.v3.protocols.column/eval13367$fn$G (column.clj:12).
No implementation of method: :to-double-array of protocol: #'tech.v3.protocols.column/PColumn found for class: nil
```

Chris has pointed out that that is `tech.v3` complaining about a missing column.

The stack trace is:
```
[09:21:29.448] [nREPL-session-1a18c2a4-9185-49a5-9193-d9406418cd6b] I pppmap.core - pppmap progress --  map: evaluate pipelines  :  0  /  0
Syntax error macroexpanding at (NO_SOURCE_FILE:38:3).
        at clojure.lang.Compiler$InvokeExpr.eval(Compiler.java:3719)
        at clojure.lang.Compiler$DefExpr.eval(Compiler.java:457)
        at clojure.lang.Compiler.eval(Compiler.java:7199)
        at clojure.lang.Compiler.eval(Compiler.java:7149)
        at clojure.core$eval.invokeStatic(core.clj:3215)
        at clojure.core$eval.invoke(core.clj:3211)
        at nrepl.middleware.interruptible_eval$evaluate$fn__1245$fn__1246.invoke(interruptible_eval.clj:87)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$with_bindings_STAR_.invokeStatic(core.clj:1990)
        at clojure.core$with_bindings_STAR_.doInvoke(core.clj:1990)
        at clojure.lang.RestFn.invoke(RestFn.java:425)
        at nrepl.middleware.interruptible_eval$evaluate$fn__1245.invoke(interruptible_eval.clj:87)
        at clojure.main$repl$read_eval_print__9206$fn__9209.invoke(main.clj:437)
        at clojure.main$repl$read_eval_print__9206.invoke(main.clj:437)
        at clojure.main$repl$fn__9215.invoke(main.clj:458)
        at clojure.main$repl.invokeStatic(main.clj:458)
        at clojure.main$repl.doInvoke(main.clj:368)
        at clojure.lang.RestFn.invoke(RestFn.java:1523)
        at nrepl.middleware.interruptible_eval$evaluate.invokeStatic(interruptible_eval.clj:84)
        at nrepl.middleware.interruptible_eval$evaluate.invoke(interruptible_eval.clj:56)
        at nrepl.middleware.interruptible_eval$interruptible_eval$fn__1278$fn__1282.invoke(interruptible_eval.clj:152)
        at clojure.lang.AFn.run(AFn.java:22)
        at nrepl.middleware.session$session_exec$main_loop__1348$fn__1352.invoke(session.clj:218)
        at nrepl.middleware.session$session_exec$main_loop__1348.invoke(session.clj:217)
        at clojure.lang.AFn.run(AFn.java:22)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalArgumentException: No implementation of method: :to-double-array of protocol: #'tech.v3.protocols.column/PColumn found for class: nil
        at clojure.core$_cache_protocol_fn.invokeStatic(core_deftype.clj:584)
        at clojure.core$_cache_protocol_fn.invoke(core_deftype.clj:576)
        at tech.v3.protocols.column$eval13367$fn__13492$G__13344__13499.invoke(column.clj:12)
        at tech.v3.dataset.column$to_double_array.invokeStatic(column.clj:201)
        at tech.v3.dataset.column$to_double_array.doInvoke(column.clj:196)
        at clojure.lang.RestFn.invoke(RestFn.java:410)
        at scicloj.metamorph.ml$supervised_eval_pipe.invokeStatic(ml.clj:45)
        at scicloj.metamorph.ml$supervised_eval_pipe.invoke(ml.clj:24)
        at scicloj.metamorph.ml$eval_pipe.invokeStatic(ml.clj:73)
        at scicloj.metamorph.ml$eval_pipe.invoke(ml.clj:65)
        at scicloj.metamorph.ml$calc_metric.invokeStatic(ml.clj:87)
        at scicloj.metamorph.ml$calc_metric.invoke(ml.clj:78)
        at scicloj.metamorph.ml$evaluate_one_pipeline$iter__37499__37503$fn__37504.invoke(ml.clj:168)
        at clojure.lang.LazySeq.sval(LazySeq.java:42)
        at clojure.lang.LazySeq.seq(LazySeq.java:51)
        at clojure.lang.RT.seq(RT.java:535)
        at clojure.core$seq__5467.invokeStatic(core.clj:139)
        at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:24)
        at clojure.core.protocols$fn__8236.invokeStatic(protocols.clj:75)
        at clojure.core.protocols$fn__8236.invoke(protocols.clj:75)
        at clojure.core.protocols$fn__8178$G__8173__8191.invoke(protocols.clj:13)
        at clojure.core$reduce.invokeStatic(core.clj:6886)
        at clojure.core$mapv.invokeStatic(core.clj:6970)
        at clojure.core$mapv.invoke(core.clj:6970)
        at scicloj.metamorph.ml$evaluate_one_pipeline.invokeStatic(ml.clj:184)
        at scicloj.metamorph.ml$evaluate_one_pipeline.invoke(ml.clj:153)
        at scicloj.metamorph.ml$evaluate_pipelines$fn__37532.invoke(ml.clj:420)
        at pppmap.core$map_with_progress$fn__29959.invoke(core.clj:66)
        at clojure.core$map_indexed$mapi__8638$fn__8639.invoke(core.clj:7376)
        at clojure.lang.LazySeq.sval(LazySeq.java:42)
        at clojure.lang.LazySeq.seq(LazySeq.java:51)
        at clojure.lang.RT.seq(RT.java:535)
        at clojure.core$seq__5467.invokeStatic(core.clj:139)
        at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:24)
        at clojure.core.protocols$fn__8236.invokeStatic(protocols.clj:75)
        at clojure.core.protocols$fn__8236.invoke(protocols.clj:75)
        at clojure.core.protocols$fn__8178$G__8173__8191.invoke(protocols.clj:13)
        at clojure.core$reduce.invokeStatic(core.clj:6886)
        at clojure.core$mapv.invokeStatic(core.clj:6970)
        at clojure.core$mapv.invoke(core.clj:6970)
        at scicloj.metamorph.ml$evaluate_pipelines.invokeStatic(ml.clj:430)
        at scicloj.metamorph.ml$evaluate_pipelines.invoke(ml.clj:272)
        at clojure.lang.AFn.applyToHelper(AFn.java:171)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$apply.invoke(core.clj:662)
        at malli.core$_instrument$fn__34038.doInvoke(core.cljc:2440)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$apply.invoke(core.clj:662)
        at malli.core$_instrument$fn__34026.doInvoke(core.cljc:2457)
        at clojure.lang.RestFn.invoke(RestFn.java:482)
        at scicloj.metamorph.ml$evaluate_pipelines.invokeStatic(ml.clj:453)
        at scicloj.metamorph.ml$evaluate_pipelines.invoke(ml.clj:272)
        at clojure.lang.AFn.applyToHelper(AFn.java:165)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$apply.invoke(core.clj:662)
        at malli.core$_instrument$fn__34038.doInvoke(core.cljc:2440)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$apply.invoke(core.clj:662)
        at malli.core$_instrument$fn__34026.doInvoke(core.cljc:2457)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.lang.Compiler$InvokeExpr.eval(Compiler.java:3714)
        ... 27 more
```

#### Further information

Taking another step into this problem...

`(cf/prediction (:metamorph/data predicted-ctx))` 
on line 31 of [scicloj.metamorph.ml](https://github.com/scicloj/metamorph.ml/blob/main/src/scicloj/metamorph/ml.clj) 
returns `nil` 
when the `:model-type` is `:sklearn.classification/random-forest-classifier`.

I'm thinking that the sklearn workings are failing to mark the inference target column.
But this is all well beyond my understanding of how the ML libs are supposed to be *fitted into* `metamorph.ml`.

#### Now

Carsten suggested that I open this issue in GitHub.

**Comment by behrica** [comment]
> I'm thinking that the sklearn workings are failing to mark the inference target column. But this is all well beyond my understanding of how the ML libs are supposed to be _fitted into_ `metamorph.ml`.
> 

Exactly that is the problem.
`evaluate-pipelines` assumes that prediction columns are "marked" as such.

All `smile` models do this, but sklearn-clj models do not do it yet.

Normal train/predict works anyway, just that the metadata is missing and only `evaluate-pipelines` relies on it.

**Comment by behrica** [comment]
Can you retry with this commit of sklearn-clj in your `deps.edn` `:git/sha cce7e8f44cec7f0a16beb9a89d2c26daa69f897c`

If It works, I make a new release of `scicloj.ml`

**Comment by ash-mcc** [comment]
Using your new commit results in:
```
java.lang.Exception: Column :species does not appear to have either a categorical or one hot map
```

Full stack trace:
```
[2022-08-31T18:53:24,829] [nREPL-session-c70ace75-2f7d-486a-976a-f45d00a512a2] I pppmap.core - pppmap progress --  map: evaluate pipelines  :  0  /  0
Syntax error macroexpanding at (NO_SOURCE_FILE:75:6).
        at clojure.lang.Compiler$InvokeExpr.eval(Compiler.java:3719)
        at clojure.lang.Compiler$InvokeExpr.eval(Compiler.java:3713)
        at clojure.lang.Compiler$DefExpr.eval(Compiler.java:457)
        at clojure.lang.Compiler.eval(Compiler.java:7199)
        at clojure.lang.Compiler.eval(Compiler.java:7149)
        at clojure.core$eval.invokeStatic(core.clj:3215)
        at clojure.core$eval.invoke(core.clj:3211)
        at nrepl.middleware.interruptible_eval$evaluate$fn__1265$fn__1266.invoke(interruptible_eval.clj:87)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$with_bindings_STAR_.invokeStatic(core.clj:1990)
        at clojure.core$with_bindings_STAR_.doInvoke(core.clj:1990)
        at clojure.lang.RestFn.invoke(RestFn.java:425)
        at nrepl.middleware.interruptible_eval$evaluate$fn__1265.invoke(interruptible_eval.clj:87)
        at clojure.main$repl$read_eval_print__9206$fn__9209.invoke(main.clj:437)
        at clojure.main$repl$read_eval_print__9206.invoke(main.clj:437)
        at clojure.main$repl$fn__9215.invoke(main.clj:458)
        at clojure.main$repl.invokeStatic(main.clj:458)
        at clojure.main$repl.doInvoke(main.clj:368)
        at clojure.lang.RestFn.invoke(RestFn.java:1523)
        at nrepl.middleware.interruptible_eval$evaluate.invokeStatic(interruptible_eval.clj:84)
        at nrepl.middleware.interruptible_eval$evaluate.invoke(interruptible_eval.clj:56)
        at nrepl.middleware.interruptible_eval$interruptible_eval$fn__1298$fn__1302.invoke(interruptible_eval.clj:152)
        at clojure.lang.AFn.run(AFn.java:22)
        at nrepl.middleware.session$session_exec$main_loop__1368$fn__1372.invoke(session.clj:218)
        at nrepl.middleware.session$session_exec$main_loop__1368.invoke(session.clj:217)
        at clojure.lang.AFn.run(AFn.java:22)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.Exception: Column :species does not appear to have either a categorical or one hot map
        at tech.v3.dataset.modelling$column_values__GT_categorical.invokeStatic(modelling.clj:123)
        at tech.v3.dataset.modelling$column_values__GT_categorical.invoke(modelling.clj:107)
        at clojure.lang.AFn.applyToHelper(AFn.java:156)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.lang.Compiler$InvokeExpr.eval(Compiler.java:3714)
        ... 28 more
```

Again, I'm just using this code from the [metamorph.ml README file](https://github.com/scicloj/metamorph.ml) but specifying the sklearn classifier instead of the SMILE classifier...
```
  (require
   '[tech.v3.dataset :as ds]
   '[tech.v3.dataset.metamorph :as ds-mm]
   '[scicloj.metamorph.core :as morph]
   '[tech.v3.dataset.modelling :as ds-mod]
   '[tech.v3.dataset.column-filters :as cf]
   '[tablecloth.api.split :as split]
   '[scicloj.metamorph.ml :as ml]
   '[scicloj.metamorph.ml.loss :as loss]
   '[scicloj.ml.smile.classification]
   '[scicloj.ml.xgboost]
   '[scicloj.sklearn-clj.ml])

  ;;  the data
  (def  ds (ds/->dataset "https://raw.githubusercontent.com/techascent/tech.ml/master/test/data/iris.csv" {:key-fn keyword}))

  ;;  the (single, fixed) pipe-fn
  (def pipe-fn
    (morph/pipeline
     (ds-mm/set-inference-target :species)
     (ds-mm/categorical->number cf/categorical)
     ;; sets the ground truth for the prediction into the context at the required key
     #_(fn [ctx]
       (assoc ctx
              :scicloj.metamorph.ml/target-ds (cf/target (:metamorph/data ctx))))
     {:metamorph/id :model}
     (ml/model {:model-type 
                ;:smile.classification/random-forest
                ;:xgboost/classification 
                :sklearn.classification/random-forest-classifier 
                })))

  ;;  the simplest split, produces a seq of one, a single split into train/test
  (def  train-split-seq (split/split->seq ds :holdout))

  ;; we have only one pipe-fn here
  (def  pipe-fn-seq [pipe-fn])

  (def  evaluations
    (ml/evaluate-pipelines pipe-fn-seq train-split-seq loss/classification-loss :loss))

  ;; we have only one result
  (def best-fitted-context (-> evaluations first first :fit-ctx))
  (def best-pipe-fn (-> evaluations first first :pipe-fn))


  ;;  simulate new data
  (def  new-ds (ds/sample ds 10 {:seed 1234}))

  ;;  do prediction on new data
  (def  predictions
    (->
     (best-pipe-fn
      (merge best-fitted-context
             {:metamorph/data new-ds
              :metamorph/mode :transform}))
     (:metamorph/data)
     (ds-mod/column-values->categorical :species)
     seq))
```

**Comment by behrica** [comment]
yes, this is an other problem.

I have not tried sklearn-clj models a lot with categorical columns.

**Comment by behrica** [comment]
I does work, when you do this:
```
(def  predictions
  (->
   (best-pipe-fn
    (merge best-fitted-context
           {:metamorph/data new-ds
            :metamorph/mode :transform}))
   (:metamorph/data)
   ;; (ds-mod/column-values->categorical :species)
   seq))
```

But then you get numbers as predictions.
You have mappings in the input dataset (as column metadata) and dataset has some functions to convert then to the the levels.

But it would indeed be "nice", if this is automatic.

(it is with Smile models, I agree)

**Comment by behrica** [comment]
I need to think a bit about this.
The issue is, that sklearn (python) does not handle at all "categoricals"  teh way I use it.
It sees only numbers, so it "predicts" only numbers.

So sklearn-clj need to keep track of the mapping in some form and re-apply it at the end.

**Comment by ash-mcc** [comment]
Carsten, thanks for the "numbers instead of labels" workaround. 
For my current purpose, I can code for that workaround.
But yes, it'd be nicer if the SMILE, XGBoost and sklearn models were totally interchangeable in the metamorph pipeline. :-)
Thank you.

**Comment by behrica** [comment]
yes, I agree with you about the interchangeability.
I will close this issue, and open a new one in sklearn-clj
